controllers/host_maintenance_controller.go

```diff
@@ -71,11 +196,215 @@ func (r *HMCMaintenanceReconciler) Reconcile(ctx context.Context, request ctrl.R
+	hostEmpty, err := r.isHostEmpty(ctx, hostSystemID, log)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if host is empty: %w", err)
+	}
+
+	if !hostEmpty {
+		log.Info("Host not empty, evacuation blocked", "host", hostSystemID)
+		return false, nil
+	}
+
+	// Gate 2: Check per-WLC ready-op-<uuid> tags on CP VMs
+	wlcReady, err := r.checkWLCReadyTags(ctx, hostSystemID, log)
+	if err != nil {
+		return false, fmt.Errorf("failed to check WLC ready tags: %w", err)
+	}
+
+	if !wlcReady {
+		log.Info("WLC ready tags not met, evacuation blocked", "host", hostSystemID)
+		return false, nil
+	}
+
+	log.Info("All evacuation gates met", "host", hostSystemID)
+	return true, nil
+}
+
+// isHostEmpty checks if the host has no running VMs
+func (r *HMCMaintenanceReconciler) isHostEmpty(ctx context.Context, hostSystemID string, log logr.Logger) (bool, error) {
+	vms, err := r.inventoryService.ListHostVMs(hostSystemID)
+	if err != nil {
+		return false, fmt.Errorf("failed to list host VMs: %w", err)
+	}
+
+	// Check if any VMs are in running state
+	for _, vm := range vms {
+		if vm.PowerState == "on" || vm.PowerState == "running" {
+			log.Info("Host has running VM", "host", hostSystemID, "vm", vm.SystemID, "powerState", vm.PowerState)
+			return false, nil
+		}
+	}
+
+	log.Info("Host is empty", "host", hostSystemID, "vmCount", len(vms))
+	return true, nil
+}
+
+// checkWLCReadyTags checks if all WLC clusters have ready-op-<uuid> tags
+func (r *HMCMaintenanceReconciler) checkWLCReadyTags(ctx context.Context, hostSystemID string, log logr.Logger) (bool, error) {
+	// Get all VMs on the host
+	vms, err := r.inventoryService.ListHostVMs(hostSystemID)
+	if err != nil {
+		return false, fmt.Errorf("failed to list host VMs: %w", err)
+	}
+
+	// Check each VM for ready-op-<uuid> tags
+	for _, vm := range vms {
+		vmDetails, err := r.inventoryService.GetVM(vm.SystemID)
+		if err != nil {
+			log.Error(err, "failed to get VM details", "vm", vm.SystemID)
+			continue
+		}
+
+		// Check if VM has any ready-op-<uuid> tags
+		hasReadyTag := false
+		for _, tag := range vmDetails.Tags {
+			if len(tag) > len(maint.TagHostOpPrefix) && tag[:len(maint.TagHostOpPrefix)] == maint.TagHostOpPrefix {
+				hasReadyTag = true
+				break
+			}
+		}
+
+		if !hasReadyTag {
+			log.Info("VM missing ready-op tag", "vm", vm.SystemID, "tags", vmDetails.Tags)
+			return false, nil
+		}
+	}
+
+	log.Info("All VMs have ready-op tags", "host", hostSystemID)
+	return true, nil
+}
+
+// clearMaintenanceTags clears maintenance-related tags from the host
+func (r *HMCMaintenanceReconciler) clearMaintenanceTags(ctx context.Context, maasMachine *infrav1beta1.MaasMachine, log logr.Logger) error {
+	if maasMachine.Spec.SystemID == nil {
+		return fmt.Errorf("maasMachine has no systemID")
+	}
+	hostSystemID := *maasMachine.Spec.SystemID
+
+	// Clear maintenance tags
+	maintenanceTags := []string{
+		maint.TagHostMaintenance,
+		maint.TagHostNoSchedule,
+	}
```

Comment:
we need 2 types for claenup for maintainance tags.

1. from HPC cluster BM is getting released, we do cleanup of HCP tags (assuming WLC tags are cleaned up by VEC reconciler)
2. When BM is freshly got deployed but have stale tags, Identify stale tags and remove them (hcp+vec tags)

Reply:
Resolved. In tags cleanup code, added logic for wtc tag cleanup also